### ISSUE TITLE
About the randomness(reproducibility) of spmm.

### ISSUE THREAD

**Issue by JiangTanZJU** [rusty1s/pytorch_sparse#82]
Hello, I use **spmm** in the traning process and I fix all the seeds like this:
~~~py
args.seed = 0
os.environ['PYTHONHASHSEED'] =str(args.seed)
random.seed(args.seed)
np.random.seed(args.seed)
torch.manual_seed(args.seed)
torch.cuda.manual_seed(args.seed)
torch.cuda.manual_seed_all(args.seed)
torch.backends.cudnn.deterministic =True
~~~

However, the loss is different among multiple exactly same training process. And if I use **torch.sparse.mm**, the problem will not occur.

For some reason, I prefer to use **spmm**. So what should I do to keep loss consistent to get the same model parameter while using spmm?

**Comment by rusty1s** [comment]
Thanks for this issue. It is correct that `spmm` is non-deterministic since it makes use of `scatter` operations under the hood. For a more advanced and deterministic sparse-matrix multiplication implementation, please have a look at `torch_sparse.SparseTensor` (still undocumented):
```python
mat1 = SparseTensor(row=row, col=col, value=value, sparse_sizes(N, M)
mat2 = torch.randn(M, K)
out = mat1 @ mat2
```
